flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamOperatorWrapper.java

```diff
@@ -0,0 +1,230 @@
+	private StreamOperatorWrapper<?, ?> next;
+
+	StreamOperatorWrapper(OP wrapped, MailboxExecutor mailboxExecutor) {
+		this.wrapped = wrapped;
+		this.mailboxExecutor = checkNotNull(mailboxExecutor);
+	}
+
+	public void close(StreamTask<?, ?> containingTask, boolean transitive) throws Exception {
+		close(containingTask, transitive, false);
+	}
+
+	/**
+	 * Ends an input of the operator contained by this wrapper.
+	 *
+	 * @param inputId the input ID starts from 1 which indicates the first input.
+	 */
+	public void endOperatorInput(int inputId) throws Exception {
+		if (wrapped instanceof BoundedOneInput) {
+			((BoundedOneInput) wrapped).endInput();
+		} else if (wrapped instanceof BoundedMultiInput) {
+			((BoundedMultiInput) wrapped).endInput(inputId);
+		}
+	}
+
+	public OP getStreamOperator() {
+		return wrapped;
+	}
+
+	void setPrevious(StreamOperatorWrapper previous) {
+		this.previous = previous;
+	}
+
+	void setNext(StreamOperatorWrapper next) {
+		this.next = next;
+	}
+
+	@VisibleForTesting
+	void close(StreamTask<?, ?> containingTask, boolean transitive, boolean invokingEndInput) throws Exception {
+		if (invokingEndInput) {
+			// NOTE: This only do for the case where the operator is one-input operator. At present,
+			// any non-head operator on the operator chain is one-input operator.
+			synchronized (containingTask.getCheckpointLock()) {
+				endOperatorInput(1);
+			}
+		}
+
+		closeOperatorAndQuiesceTimeService(containingTask);
+
+		// transmit the close operation to the next operator on the operator chain in the topological order
+		if (transitive && next != null) {
+			next.close(containingTask, true, true);
+		}
+	}
+
+	private void closeOperatorAndQuiesceTimeService(StreamTask<?, ?> containingTask)
+		throws InterruptedException, ExecutionException {
+
+		final CompletableFuture<Void> closedFuture = new CompletableFuture<>();
+
+		// 1. executing the close operation must be deferred to the mailbox to ensure that mails already
+		//    in the mailbox are finished before closing the operator
+		// 2. to ensure that there is no longer output triggered by the timers before invoke the "endInput"
+		//    methods of downstream operators in the operator chain, we must quiesce the processing time
+		//    service to prevent the pending timers from firing, but allow the timers in running to finish
+		// 3. when the second step is finished, send a close mail to ensure that the timers yet in mailbox
+		//    are finished before exiting the following mailbox processing loop
+		// TODO: To ensure the strict semantics of "close", the operator should be allowed to decide how to
+		//  handle (cancel or trigger) the pending timers before being closed, and the second step should be
+		//  finished before the first step
+		deferCloseOperatorToMailbox(containingTask)
+			.thenRun(this::quiesceProcessingTimeService)
+			.thenRun(() -> sendClosedMail(closedFuture))
+			.exceptionally((throwable) -> forwardException(throwable, closedFuture));
```

Comment:
This could be simplified if we'd create future inside send and then compose it:
```
final CompletableFuture<Void> closedFuture =
    deferCloseOperatorToMailbox(containingTask)
        .thenRun(this::quiesceProcessingTimeService)
        .thenCompose(unused -> sendClosedMail());
...
private CompletableFuture<Void> sendClosedMail() {
    CompletableFuture<Void> f = new CompletableFuture<>();
    mailboxExecutor.execute(() -> f.complete(null), "StreamOperatorWrapper#sendClosedMail for " + wrapped);
    return f;
}
```
`forwardException` method then could be dropped.

Reply:
After testing, using `thenCompose` can cause a deadlock in the mailbox processing loop. Because setting `closedFuture`  to the `done` state may be after `yield()`, which causes `yield()` to be called again, but at this time, all mails of the operator have been processed.